github/changedProjects.go

```diff
@@ -0,0 +1,69 @@
+package github
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+)
+
+func appendIfMissing(slice []string, s string) []string {
+	for _, ele := range slice {
+		if ele == s {
+			return slice
+		}
+	}
+	return append(slice, s)
+}
+
+func getTopLevelDirName(path string) string {
+	const separator = "/"
+	if !strings.Contains(path, separator) {
+		return ""
+	}
+	return strings.Split(path, separator)[0]
+}
+
+func getTree(repo *git.Repository, ref string) (tree *object.Tree, err error) {
+	hash := plumbing.NewHash(ref)
+	commit, err := repo.CommitObject(hash)
+	if err != nil {
+		err = fmt.Errorf("Cannot get commit from hash %v: %v", hash, err)
+		return
+	}
+	tree, err = commit.Tree()
+	if err != nil {
+		err = fmt.Errorf("Cannot get tree from commit %v: %v", commit, err)
+		return
+	}
+	return
+}
+
+// GetChangedProjects returns an array of unique top level directory names
+// in which there have been changes
+func GetChangedProjects(repo *git.Repository, headRef, baseRef string) (files []string, err error) {
+	headTree, err := getTree(repo, headRef)
+	if err != nil {
+		err = fmt.Errorf("Cannot get tree from commit %v: %v", headRef, err)
+		return
+	}
+	baseTree, err := getTree(repo, baseRef)
+	if err != nil {
+		err = fmt.Errorf("Cannot get tree from commit %v: %v", headRef, err)
+		return
+	}
+	diff, err := headTree.Diff(baseTree)
+	for _, change := range diff {
+		name := change.To.Name
+		if name == "" {
+			name = change.From.Name
+		}
+		dir := getTopLevelDirName(name)
+		if dir != "" {
+			files = appendIfMissing(files, dir)
```

Comment:
Not a biggie, but would it be nicer to collect the names into a set? You can use a map for that kind of thing:

```
set := make(map[int]bool)
```

Then just loop over it to get the slice of unique names.

Reply:
Totally get your point but I like the symmetry with `append(slice, element)` and as we're working with slices anyway, I think this is simpler and more intuitive. The map may well perform better, but the function does exit early and we're only ever going to be looping round the changed top level dirs in the repo, so not big numbers of elements.

Reply:
@charypar thanks - have refactored.